src/main/java/com/iota/iri/network/pipeline/TransactionProcessingPipeline.java

```diff
@@ -0,0 +1,356 @@
+    private BatchedHasher batchedHasher;
+    private HashingStage hashingStage;
+
+    private FIFOCache<Long, Hash> recentlySeenBytesCache;
+    private BlockingQueue<ProcessingContext<PreProcessPayload>> preProcessStageQueue = new ArrayBlockingQueue<>(100);
+    private BlockingQueue<
+            ProcessingContext<? extends ValidationPayload>> validationStageQueue = new ArrayBlockingQueue<>(100);
+    private BlockingQueue<ProcessingContext<ReceivedPayload>> receivedStageQueue = new ArrayBlockingQueue<>(100);
+    private BlockingQueue<ProcessingContext<BroadcastPayload>> broadcastStageQueue = new ArrayBlockingQueue<>(100);
+    private BlockingQueue<ProcessingContext<ReplyPayload>> replyStageQueue = new ArrayBlockingQueue<>(100);
+
+    /**
+     * Defines the different stages of the {@link TransactionProcessingPipeline}.
+     */
+    public enum Stage {
+        PRE_PROCESS, HASHING, VALIDATION, REPLY, RECEIVED, BROADCAST, MULTIPLE, ABORT
+    }
+
+    /**
+     * Initializes the dependencies of the {@link TransactionProcessingPipeline}.
+     * 
+     * @param neighborRouter         The {@link NeighborRouter} to use for broadcasting transactions
+     * @param config                 The config to set cache sizes and other options
+     * @param txValidator            The transaction validator to validate incoming transactions with
+     * @param tangle                 The {@link Tangle} database to use to store and load transactions.
+     * @param snapshotProvider       The {@link SnapshotProvider} to use to store transactions with.
+     * @param tipsViewModel          The {@link TipsViewModel} to load tips from in the reply stage
+     * @param latestMilestoneTracker The {@link LatestMilestoneTracker} to load the latest milestone hash from in the
+     *                               reply stage
+     */
+    public void init(NeighborRouter neighborRouter, NodeConfig config, TransactionValidator txValidator, Tangle tangle,
+            SnapshotProvider snapshotProvider, TipsViewModel tipsViewModel,
+            LatestMilestoneTracker latestMilestoneTracker) {
+        this.recentlySeenBytesCache = new FIFOCache<>(config.getCacheSizeBytes());
+        this.preProcessStage = new PreProcessStage(recentlySeenBytesCache);
+        this.replyStage = new ReplyStage(neighborRouter, config, tangle, tipsViewModel, latestMilestoneTracker,
+                snapshotProvider, recentlySeenBytesCache);
+        this.broadcastStage = new BroadcastStage(neighborRouter);
+        this.validationStage = new ValidationStage(txValidator, recentlySeenBytesCache);
+        this.receivedStage = new ReceivedStage(tangle, txValidator, snapshotProvider);
+        this.batchedHasher = BatchedHasherFactory.create(BatchedHasherFactory.Type.BCTCURL81, 20);
+        this.hashingStage = new HashingStage(batchedHasher);
+    }
+
+    /**
+     * Kicks of the pipeline by spawning all necessary threads for all stages.
+     */
+    public void start() {
```

Comment:
Im having issues with this pipeline :D Rather see it split.

If the stage has an interface, we can do this: https://pastebin.com/PSG5hyKG
Have to change the ctx return a bit. Although then were back at how the API was -> functional interfaces? :D
Otherwise, we can do this:
```
addStage("broadcast", () -> {
            try {
                broadcastStage();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
    }

    private void addStage(String name, Runnable object) {
        stagesThreadPool.submit(new Thread(() -> {
            try {
                while (!shutdown.get()) {
                    object.run();
                }
            } catch (RuntimeException e) {
                
            } finally {
                log.info("{} stage shutdown", name);
            }
        }, name + "-stage"));
    }
```

Idk, gotta stop thinking about it now.
It does the job, so... :shrug:

Reply:
Cool idea. Added addStage(), pipeline assembly looks cleaner now.